h2o-core/src/main/java/water/util/VecUtils.java

```diff
@@ -879,4 +882,93 @@ public void map(Chunk c, NewChunk nc) {
       result += mrt.result;
     }
   }
+
+  /**
+   * Randomly shuffle a Vec using Fisher Yates shuffle 
+   * https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
+   */
+  public static class ShuffleVecTask extends MRTask<ShuffleVecTask> {
+    private final long _seed;
+    private final Vec _vec;
+    private transient int[] _localChunks;
+    private transient int[] _permutatedChunks;
+
+    public ShuffleVecTask(final Vec vec, final long seed) {
+      super();
+      _seed = seed;
+      _vec = vec;
+    }
+
+    @Override
+    protected void setupLocal() {
+      _localChunks = VecUtils.getLocalChunkIds(_vec);
+      _permutatedChunks = _localChunks.clone();
+      permute(_permutatedChunks, null);
+    }
+
+    private void permute(int[] arr, Random rng) {
+      if (null == rng) rng = getRNG(_seed);
+      for (int i = arr.length - 1; i > 0; i--) {
+        int j = rng.nextInt(i + 1);
+        final int old = arr[i];
+        arr[i] = arr[j];
+        arr[j] = old;
+      }
+    }
+
+    @Override
+    public void map(Chunk _cs, NewChunk nc) {
+      Random rng = getRNG(_seed + _cs.start());
+      Chunk cs = _vec.chunkForChunkIdx(_permutatedChunks[Arrays.binarySearch(_localChunks, _cs.cidx())]);
+
+      int[] permutedRows = IntStream.range(0, cs._len).toArray();
+      permute(permutedRows, rng);
+
+      for (int row : permutedRows) {
+        if (cs.isNA(row)) {
+          nc.addNA();
+        } else {
+          switch (_cs.vec().get_type()) {
+            case Vec.T_BAD:
+              break;
+            case Vec.T_UUID:
+              nc.addUUID(cs, row);
```

Comment:
random access to chunk elements is expensive, it might be cheaper to extract to a dense array - shuffle and write to nc

- just a thought

Reply:
Thank you, I briefly looked in to it and it seems to me that it would look something like the following pseudo-code :
```java
if (c.hasNA()) {
  for (int r=0; r < c._len; r++)
    if (!c.isNA(r)) arr[r] = c.at*(r);
} else {
    for (int r=0; r < c._len; r++)
      arr[r] = c.at*(r);
}

arr_indices = 0 ... arr.length;
permute(arr_indices, seed);
permute(arr, seed);

if (c.hasNA()) {
  for (int r=0; r < c._len; r++)
    if (c.isNA(arr_indices[r])) 
      nc.addNA();
    else
      nc.add*(arr[r]);
} else {
  for (int r=0; r < c._len; r++)
    nc.add*(arr[r]);
}
```

If that's so I think I will still be in need of doing random access to check if it was a missing value (unless it was a double). Anyway, I will look into it more tomorrow, maybe I'll find a better way.

Reply:
I would probably not worry about it in the context of the varimp-calculation. Most of the complexity will be in scoring most likely.

This seems good as-is.